### PR TITLE
Remove `omp_get_max_threads`

### PR DESCRIPTION
**Pull request by trivialfis**
This is the one last PR for removing omp global variable.

* Add context object to the `DMatrix`.  This bridges `DMatrix` with https://github.com/dmlc/xgboost/issues/7308 .
* Require context to be available at the construction time of booster.
* Add `n_threads` support for R csc DMatrix constructor.
* Remove `omp_get_max_threads` in R glue code.
* Remove threading utilities that rely on omp global variable.

Close https://github.com/dmlc/xgboost/issues/7583 .

### CODE REVIEW

File: src/learner.cc [hunk 1]
```diff
@@ -331,7 +331,6 @@ class LearnerConfiguration : public Learner {
     generic_parameters_.UpdateAllowUnknown(args);
 
     ConsoleLogger::Configure(args);
-    common::OmpSetNumThreads(&generic_parameters_.nthread);
```
Comment (trivialfis) [inline src/learner.cc:334]: From now on XGBoost will no longer mutate the omp global variable.

**Comment by trivialfis** [comment]
Tried to eliminate omp parallel clause, but apple clang is not quite happy about the use of lambda function.

**Review by hcho3 — COMMENTED** [review]
Overall LGTM. Who owns the `ctx` object? I hope we don't end up with a dangling object, since many components take a pointer to `ctx` object that's owned by someone else.

**Comment by trivialfis** [comment]
@hcho3 Both the learner and DMatrix own a ctx. They are the first to be initialized and the last to be destroyed.

**Pull request merged**